File: go/fiveproblems2.go
// Write a function that combines two lists by alternatingly taking elements.
// For example: given the two lists [a, b, c] and [1, 2, 3], the function should
// return [a, 1, b, 2, c, 3].
package main

import "fmt"

func alternate(l1 []string, l2 []string) (result []string) {
    size := len(l1)
    if len(l2) > size {
        size = len(l2)
    }
    for i := 0; i < size; i++ {
        if i < len(l1) {
            result = append(result, l1[i])
        }
        if i < len(l2) {
            result = append(result, l2[i])
        }
    }
    return
}

func main() {
    v1 := []string{"A", "B", "C", "D"}
    v2 := []string {"1", "2", "3", "4", "XX", "YY"}
    fmt.Println(alternate(v1, v2))
    fmt.Println(alternate(v2, v1))
}

File: rs/fiveproblems2.rs
// Write a function that combines two lists by alternatingly taking elements.
// For example: given the two lists [a, b, c] and [1, 2, 3], the function should
// return [a, 1, b, 2, c, 3].

fn main() {
    let v1: Vec<String> =
        vec!["A".to_string(), "B".into(), "C".into(), "D".into()];
    let v2: Vec<String> =
        vec!["1".to_string(), "2".into(), "3".into(), "4".into(), "XX".into(),
            "YY".into()];
    let result = alternate(v1, v2);
    println!("{:?}", result);
}

fn alternate(v1: Vec<String>, v2: Vec<String>) -> Vec<String> {
    let mut result: Vec<String> = Vec::new();
    let mut i = 0;
    while i < v1.len() || i < v2.len() {
        if i < v1.len() {
            result.push(v1[i].clone());
        }
        if i < v2.len() {
            result.push(v2[i].clone());
        }
        i += 1;
    }
    result
}

